wflambda.go

```diff
@@ -0,0 +1,283 @@
+package wflambda
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-lambda-go/lambdacontext"
+	"github.com/rcrowley/go-metrics"
+	"github.com/wavefronthq/go-metrics-wavefront"
+	"log"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+)
+
+type lambdaHandlerFunction func(context.Context, json.RawMessage) (interface{}, error)
+
+var (
+	server                  string
+	authToken               string
+	isReportStandardMetrics bool
+	csCounter               metrics.Counter
+	csEventCounter          metrics.Counter
+	invocationsCounter      metrics.Counter
+	invocationEventCounter  metrics.Counter
+	eCounter                metrics.Counter
+	eEventCounter           metrics.Counter
+	dGauge                  metrics.Gauge
+	isColdStart             = true
+)
+
+/* Returns the Wavefront Lambda wrapper. The wrapper collects aws lambda's
+   standard metrics and reports it directly to the specified wavefront url. It
+   requires the following Environmental variables to be set:
+   1.WAVEFRONT_URL : https://<INSTANCE>.wavefront.com
+   2.WAVEFRONT_API_TOKEN : Wavefront API token with Direct Data Ingestion permission
+   3.IS_REPORT_STANDARD_METRICS : Set to False to not report standard lambda
+                                 metrics directly to wavefront.
+*/
+func Wrapper(lambdaHandler interface{}) interface{} {
+	handlerType := reflect.TypeOf(lambdaHandler)
+	handlerValue := reflect.ValueOf(lambdaHandler)
+	// Validate lambdaHandler Kind.
+	if handlerType.Kind() != reflect.Func {
+		return lambdaErrorHandler(fmt.Errorf("Expected lambda handler function type : %s , but Actual : %s", handlerType.Kind(), reflect.Func))
+	}
+
+	// Validate lambdaHandler Input Arguments.
+	isLambdaHandlerTakesContext := false
+	switch numberTIn := handlerType.NumIn(); numberTIn {
+	case 0: // do nothing
+	case 1:
+		cxtType := reflect.TypeOf((*context.Context)(nil)).Elem()
+		inputArgType := handlerType.In(0)
+		isLambdaHandlerTakesContext = inputArgType.Implements(cxtType)
+	case 2:
+		cxtType := reflect.TypeOf((*context.Context)(nil)).Elem()
+		inputArgType := handlerType.In(0)
+		isLambdaHandlerTakesContext = inputArgType.Implements(cxtType)
+		if !isLambdaHandlerTakesContext {
+			return lambdaErrorHandler(fmt.Errorf("There lambda handler function takes 2 arguments. Expected First Argument of to be of kind : %s , but Actual : %s", cxtType, inputArgType))
+		}
+	default:
+		return lambdaErrorHandler(fmt.Errorf("The lambda handler function takes incorrect number of arguments. It takes %d input arguments", handlerType.NumIn()))
+	}
+
+	isReportStandardMetrics = getAndValidateLambdaEnvironment()
+	if !isReportStandardMetrics {
+		// Returns a plain wrapper function without gathering standard lambda metrics.
+		return func(ctx context.Context, payload json.RawMessage) (response interface{}, lambdaHandlerError error) {
```

Comment:
Change the return signature to just `(interface{}, error)`? Currently uses named return parameters. But the actual function doesn't have any such named params.